community/rfcs/techniques/proposals/20200205-weight-clustering-proposal.md

```diff
@@ -0,0 +1,82 @@
+# Keras Weight Clustering API
+
+| Status        | Accepted                                      |
+:---------------|:----------------------------------------------|
+| **Author(s)** | Mohamed Nour Abouelseoud (mohamednour.abouelseoud@arm.com), Aron Virginas-Tar (aron.virginas-tar@arm.com), Anton Kachatkou (anton.kachatkou@arm.com)|
+| **Sponsor**   | Alan Chiao (alanchiao@google.com)             |
+| **Updated**   | 2020-03-06                                    |
+
+## Objective
+
+This RFC document addresses the technique &ndash; and consequent API change &ndash; of weight clustering. This is an approach for compressing neural networks by restricting the number of unique weights so that a lower number of bits is required to represent each weight.
+
+## Motivation
+
+Clustering, or weight sharing, allows for certain types of hardware to benefit from advanced weight compression techniques and the associated reduction in model memory footprint and bandwidth.
+Clustering can also be combined with pruning and quantization, reducing the network’s storage requirements even further.
+
+See the [experimental results](#experimental-results) below, obtained across a set of models, that demonstrate the benefits of this technique in terms of weight compression and accuracy preservation.
+
+## Proposed Design
+
+The weight clustering technique that forms the basis of our implementation has been described in the conference paper *Deep Compression: Compressing Deep Neural Networks with Pruning, Trained Quantization and Huffman Coding* by S. Han, H. Mao, and  W. J. Dally, which can be downloaded from [arXiv](https://arxiv.org/abs/1510.00149). Please refer to chapter 3, entitled *Trained Quantization and Weight Sharing*, for details about the clustering technique &mdash; since the paper is brief and freely available for download, we will not reproduce its content here.
+
+From a practical standpoint, the implementation uses a lookup table to hold the cluster centroid values during model training. The weight array is populated with a 'gather' operation so that during back propagation the gradients can be calculated in the normal way. The lookup table is then adjusted using the cumulative gradient values for the weights that correspond to the same centroid.
+
+The number of unique values required, as well as the way the cluster centroids are initialized are passed in as parameters.
+
+The initial values of cluster centroids are fine-tuned during a subsequent training step.
+
+## Experimental Results
+
+The table below summarizes the compression rate and top-1 accuracy achieved when clustering the weights of different models using linear centroid initialization:
+
+<table>
```

Comment:
Please separate into three tables for the three different tasks (like the [pruning RFC](https://github.com/tensorflow/model-optimization/pull/268/files?short_path=edfd107#diff-edfd10773bb6bb7662ce4fb0b94d12a1)).

It highlights that you have three tasks and task coverage is just as important as model coverage.